linera-chain/src/block_tracker.rs

```diff
@@ -149,14 +154,36 @@ impl<'resources> BlockExecutionTracker<'resources> {
             ))
             .with_execution_context(context)?;
 
+        // Account for blobs published by this transaction directly.
         for blob in &txn_outcome.blobs {
-            if blob.content().blob_type() == BlobType::Data {
+            resource_controller
+                .policy()
+                .check_blob_size(blob.content())
+                .with_execution_context(context)?;
+            if blob.content().blob_type().is_user_blob() {
                 resource_controller
                     .track_blob_published(blob.content())
                     .with_execution_context(context)?;
             }
         }
 
+        // Account for blobs published indirectly but referenced by the transaction.
+        for blob_id in &txn_outcome.blobs_published {
+            if let Some(blob) = self.published_blobs.get(blob_id) {
+                // We don't check its sizes again, it was check at the beginning
+                // of the block processing.
+                if blob.content().blob_type().is_user_blob() {
+                    resource_controller
+                        .track_blob_published(blob.content())
+                        .with_execution_context(context)?;
+                }
+            } else {
+                return Err(ChainError::InternalError(format!(
```

Comment:
(This is "internal error" because we already check that we have all published blobs in an earlier step, is it?)

Reply:
Yes, exactly.